### ISSUE TITLE
Is it possible to DistinctBy latest records?

### ISSUE THREAD

**Issue by MousKamel** [morelinq/MoreLINQ#504]

**Comment by atifaziz** [comment]
I'm not sure what you're asking exactly. Can you rephrase the question or expand with some details? What does a record look like? What are you after as the desired output?

**Comment by MousKamel** [comment]
I have a records that have a column that I can distinct. but when I use DistinctBy it returns me the first record of each groub. I would like to have the last records of each group.

**Comment by atifaziz** [comment]
Can you share some self-contained code to demonstrate the issue?

**Comment by leandromoh** [comment]
@MousKamel  One workaround is 

souce
.GroupBy(x => x.Property)
.SelectMany(x => x.Last())

@atifaziz maybe would be worth to add other overload where user pass a func that receives the selected element and current and return one. Similar to "Func<T, T, T> fillSelector" in FillBackward.

**Comment by atifaziz** [comment]
@leandromoh [`GroupBy` already has such an overload](https://docs.microsoft.com/en-us/dotnet/api/system.linq.enumerable.groupby?view=netframework-4.7.1#System_Linq_Enumerable_GroupBy__3_System_Collections_Generic_IEnumerable___0__System_Func___0___1__System_Func___1_System_Collections_Generic_IEnumerable___0____2__) so you don't need `SelectMany`.

Assuming you understood @MousKamel's question better than I did (was confused between _latest_ and _last_), here is a standalone example that gets distinct word lengths along with the _last_ seen word of that length:

```c#
var text = @"
    Lorem ipsum dolor sit amet, consectetur adipiscing elit.
    Etiam gravida nec mauris vitae sollicitudin. Suspendisse
    malesuada urna eu mi suscipit fringilla. Donec ut ipsum
    aliquet, tincidunt mi sed, efficitur magna. Nulla sit
    amet congue diam, at posuere lectus. Praesent sit amet
    libero vehicula dui commodo gravida eget a nisi. Sed
    imperdiet arcu eget erat feugiat gravida id non est.
    Nam malesuada nibh sit amet nisl sollicitudin vestibulum.";

var words = Regex.Matches(text, @"\w+");

var q =
    from Match m in words
    select m.Value into s
    group s by s.Length into g
    orderby g.Key
    select g.Key + ": " + g.Last();

foreach (var e in q)
    Console.WriteLine(e);
```

outputs:

```
1: a
2: id
3: sit
4: nisl
5: Nulla
6: libero
7: gravida
8: vehicula
9: malesuada
10: vestibulum
11: Suspendisse
12: sollicitudin
```

Here's the same query using the `GroupBy` overload I mentioned earlier:

```c#
var q =
    from g in words.Cast<Match>()
                   .Select(m => m.Value)
                   .GroupBy(s => s.Length,
                            (len, ms) => new { Length = len, LastWord = ms.Last() })
    orderby g.Length
    select g.Length + ": " + g.LastWord;
```

**Comment by MousKamel** [comment]
Sorry, I didnt understand your answer. Ill be more specific now. I have this table:

```
id |          date                | cote |     message |     userId                                         |     Seller | totalCote
33 | 10/05/2018 22:13:20 | 1 | <p>ll</p> | 32b23ff3-d323-4b64-b331-c6eb78a8c88f | lara | 1
34 | 10/05/2018 22:21:07 | 5 | <p>bien</p> | 32b23ff3-d323-4b64-b331-c6eb78a8c88f | lara | 6
35 | 10/05/2018 22:22:42 | 3 | <p>c va</p> | 32b23ff3-d323-4b64-b331-c6eb78a8c88f | lara | 9
36 | 10/05/2018 22:26:46 | -6 | <p>mal</p> | 1a9f207a-3a11-42c9-90db-2e0a49050022 | lara | 3
37 | 11/05/2018 23:56:09 | 5 | <p>bien</p> | a0690e57-f528-48cc-adf6-fd73df7edb6f | mara | 5
38 | 23/05/2018 12:52:14 | -1 | <p>alo</p> | 32b23ff3-d323-4b64-b331-c6eb78a8c88f | mara | 4
39 | 23/05/2018 12:53:51 | 1 | <p>sss</p> | 32b23ff3-d323-4b64-b331-c6eb78a8c88f | mara | 5
40 | 24/05/2018 08:24:45 | -3 | <p>aaa</p> | a0690e57-f528-48cc-adf6-fd73df7edb6f | lara | 0
41 | 24/05/2018 08:25:24 | -3 | <p>aaa1</p> | a0690e57-f528-48cc-adf6-fd73df7edb6f | lara | -3
42 | 24/05/2018 08:25:42 | -3 | <p>aaa2</p> | a0690e57-f528-48cc-adf6-fd73df7edb6f | lara | -6
43 | 24/05/2018 08:38:06 | -1 | <p>moi -1</p> | a0690e57-f528-48cc-adf6-fd73df7edb6f | lara | -7
44 | 24/05/2018 08:50:03 | -1 | <p>encore -1</p> | a0690e57-f528-48cc-adf6-fd73df7edb6f | lara | -8
45 | 24/05/2018 08:54:12 | 1 | <p>malo</p> | a0690e57-f528-48cc-adf6-fd73df7edb6f | lara | -7
46 | 24/05/2018 08:56:07 | 1 | <p>malo1</p> | a0690e57-f528-48cc-adf6-fd73df7edb6f | lara | -6
47 | 25/05/2018 12:33:32 | -3 | <p>alo</p> | 32b23ff3-d323-4b64-b331-c6eb78a8c88f | mara | 2
48 | 25/05/2018 12:33:42 | -3 | <p>alo</p> | 32b23ff3-d323-4b64-b331-c6eb78a8c88f | mara | -1
49 | 25/05/2018 12:33:47 | -3 | <p>alo</p> | 32b23ff3-d323-4b64-b331-c6eb78a8c88f | mara | -4
50 | 25/05/2018 12:33:54 | -3 | <p>alo</p> | 32b23ff3-d323-4b64-b331-c6eb78a8c88f | mara | -7
51 | 25/05/2018 12:34:41 | -1 | <p>alo</p> | 32b23ff3-d323-4b64-b331-c6eb78a8c88f | mara | -8
```
I know the field `totalCote` is useless since I can just group according to the Seller and sum up the `cote` but just out of curiosity, since it wasn't me who modeled this. I know that the latest record of any seller has the sum of its cotes in the field totalCote.  So by using: 

```
List<Evaluation> evaluations = new List<Evaluation>();
evaluations = db.Evaluations.DistinctBy(c => c.Vendeur).ToList();
```

the result is : 
```
33 | 10/05/2018 22:13:20 | 1 | <p>ll</p> | 32b23ff3-d323-4b64-b331-c6eb78a8c88f | lara | 1
37 | 11/05/2018 23:56:09 | 5 | <p>bien</p> | a0690e57-f528-48cc-adf6-fd73df7edb6f | mara | 5
```
but I want to know how I can get by adding a linq method this result:

```
46 | 24/05/2018 08:56:07 | 1 | <p>malo1</p> | a0690e57-f528-48cc-adf6-fd73df7edb6f | lara | -6
51 | 25/05/2018 12:34:41 | -1 | <p>alo</p> | 32b23ff3-d323-4b64-b331-c6eb78a8c88f | mara | -8
```
which are the latests record of each sellers. I hope this is clear now. Btw I know I complicate things but it's just out of curiosity. I have a simpler solution already. Thank you for your time ! :)

**Comment by PhilippeRaemy** [comment]
@MousKamel : It would appear that you have found yourself an acceptable answer. However, the approach proposed by @atifaziz gives you potentially more possibilities.
Applied to your example it would translate to:
```c#
evaluations = db.Evaluations   
                .GroupBy(c => c.Vendeur, (_, g) => g.OrderByDescending(c => c.Id).First())
```

The *keySelector* argument of the `GroupBy` lets you group by vendor, then with the *resultSelector* lambda you can extract whatever you want from the vendors's set: first record, last record, compute an average, ... and build your result to fit your need.
Note that, in this example of *resultSelector* I've used only the group members enumeration, hence leaving the *key* argument of the resultSelector as an anonymous _.

Additionally, I would say that there are good chances that, on a large list, grouping and then ordering on each group performs faster than ordering the whole list and then searching for distinct values. I don't have a literal proof of that but it's certainly worth a test.

**Comment by atifaziz** [comment]
```c#
const string data =
  @"id |          date  | cote |     message |           userId                  |   Seller | totalCote
    33 | 10/05/2018 22:13:20 | 1 | <p>ll</p> | 32b23ff3-d323-4b64-b331-c6eb78a8c88f | lara | 1
    34 | 10/05/2018 22:21:07 | 5 | <p>bien</p> | 32b23ff3-d323-4b64-b331-c6eb78a8c88f | lara | 6
    35 | 10/05/2018 22:22:42 | 3 | <p>c va</p> | 32b23ff3-d323-4b64-b331-c6eb78a8c88f | lara | 9
    36 | 10/05/2018 22:26:46 | -6 | <p>mal</p> | 1a9f207a-3a11-42c9-90db-2e0a49050022 | lara | 3
    37 | 11/05/2018 23:56:09 | 5 | <p>bien</p> | a0690e57-f528-48cc-adf6-fd73df7edb6f | mara | 5
    38 | 23/05/2018 12:52:14 | -1 | <p>alo</p> | 32b23ff3-d323-4b64-b331-c6eb78a8c88f | mara | 4
    39 | 23/05/2018 12:53:51 | 1 | <p>sss</p> | 32b23ff3-d323-4b64-b331-c6eb78a8c88f | mara | 5
    40 | 24/05/2018 08:24:45 | -3 | <p>aaa</p> | a0690e57-f528-48cc-adf6-fd73df7edb6f | lara | 0
    41 | 24/05/2018 08:25:24 | -3 | <p>aaa1</p> | a0690e57-f528-48cc-adf6-fd73df7edb6f | lara | -3
    42 | 24/05/2018 08:25:42 | -3 | <p>aaa2</p> | a0690e57-f528-48cc-adf6-fd73df7edb6f | lara | -6
    43 | 24/05/2018 08:38:06 | -1 | <p>moi -1</p> | a0690e57-f528-48cc-adf6-fd73df7edb6f | lara | -7
    44 | 24/05/2018 08:50:03 | -1 | <p>encore -1</p> | a0690e57-f528-48cc-adf6-fd73df7edb6f | lara | -8
    45 | 24/05/2018 08:54:12 | 1 | <p>malo</p> | a0690e57-f528-48cc-adf6-fd73df7edb6f | lara | -7
    46 | 24/05/2018 08:56:07 | 1 | <p>malo1</p> | a0690e57-f528-48cc-adf6-fd73df7edb6f | lara | -6
    47 | 25/05/2018 12:33:32 | -3 | <p>alo</p> | 32b23ff3-d323-4b64-b331-c6eb78a8c88f | mara | 2
    48 | 25/05/2018 12:33:42 | -3 | <p>alo</p> | 32b23ff3-d323-4b64-b331-c6eb78a8c88f | mara | -1
    49 | 25/05/2018 12:33:47 | -3 | <p>alo</p> | 32b23ff3-d323-4b64-b331-c6eb78a8c88f | mara | -4
    50 | 25/05/2018 12:33:54 | -3 | <p>alo</p> | 32b23ff3-d323-4b64-b331-c6eb78a8c88f | mara | -7
    51 | 25/05/2018 12:34:41 | -1 | <p>alo</p> | 32b23ff3-d323-4b64-b331-c6eb78a8c88f | mara | -8";

var evals =
    from line in Regex.Split(data, @"\r?\n")
                      .Select(line => line.Trim())
                      .Where(line => line.Length > 0)
                      .Skip(1) // skip header
    select line.Split('|')
               .Select(e => e.Trim())
               // below assumes column order
               .Fold((id, date, cote, msg, uid, seller, tc) => new
               {
                   Id        = int.Parse(id),
                   Date      = DateTime.ParseExact(date, "dd'/'MM'/'yyyy HH':'mm':'ss", null),
                   Cote      = cote,
                   Message   = msg,
                   UserId    = uid,
                   Seller    = seller,
                   TotalCote = tc,
               });

var lastBySeller =
    from eval in evals
 // orderby eval.Seller, eval.Date
    group eval by eval.Seller into g
    select g.Last();
```

@MousKamel The first query above just parses the data sample you shared into parsed records. The second query shows how to use `GroupBy` and `Last` to get the last record of each seller. Uncomment the `orderby` clause if the data may not be sorted and your truly want the _latest_ record, instead of just the _last_, by seller.

You can do this with standard LINQ and has nothing to do with MoreLINQ's `DistinctBy` so I am going to consider this closed this for now.
